lib/__tests__/functions.test.ts

```diff
@@ -0,0 +1,159 @@
+import fs, { PathLike } from 'fs-extra';
+import findup from 'findup-sync';
+import { getCwd } from '../path';
+import { downloadGithubRepoContents } from '../github';
+import { logger } from '../logger';
+import {
+  createFunction,
+  isObjectOrFunction,
+  createEndpoint,
+  createConfig,
+} from '../cms/functions';
+
+jest.mock('fs-extra');
+jest.mock('findup-sync');
+jest.mock('../path');
+jest.mock('../github');
+
+const mockedGetCwd = getCwd as jest.MockedFunction<typeof getCwd>;
+const mockedFindup = findup as jest.MockedFunction<typeof findup>;
+const mockedFsExistsSync = fs.existsSync as jest.MockedFunction<
+  typeof fs.existsSync
+>;
+const mockedFsReadFileSync = fs.readFileSync as jest.MockedFunction<
+  typeof fs.readFileSync
+>;
+
+describe('functions', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('createFunction', () => {
+    const mockFunctionInfo = {
+      functionsFolder: 'testFolder',
+      filename: 'testFunction',
+      endpointPath: '/api/test',
+      endpointMethod: 'GET',
+    };
+    const mockDest = '/mock/dest';
+
+    beforeEach(() => {
+      mockedGetCwd.mockReturnValue('/mock/cwd');
+      mockedFindup.mockReturnValue(null);
+
+      // Set up fs.existsSync to return different values for different paths
+      mockedFsExistsSync.mockImplementation((path: PathLike) => {
+        if (path === '/mock/dest/testFolder.functions/testFunction.js') {
+          return false;
+        }
+        if (path === '/mock/dest/testFolder.functions/serverless.json') {
+          return true;
+        }
+        if (path === '/mock/dest/testFolder.functions') {
+          return true;
+        }
+        return false;
+      });
+
+      // Mock fs.readFileSync to return a valid JSON for the config file
+      mockedFsReadFileSync.mockReturnValue(
+        JSON.stringify({
+          endpoints: {},
+        })
+      );
+    });
+
+    it('should create a new function successfully', async () => {
+      await createFunction(mockFunctionInfo, mockDest);
+
+      expect(fs.mkdirp).not.toHaveBeenCalled();
+
+      expect(downloadGithubRepoContents).toHaveBeenCalledWith(
+        'HubSpot/cms-sample-assets',
+        'functions/sample-function.js',
+        '/mock/dest/testFolder.functions/testFunction.js'
+      );
+
+      // Check that the config file was updated
+      expect(fs.writeFileSync).toHaveBeenCalledWith(
+        '/mock/dest/testFolder.functions/serverless.json',
+        expect.any(String)
+      );
+
+      // Verify that the logger was called with the correct messages
+      jest.spyOn(logger, 'log').mockImplementation(() => {
```

Comment:
IMO we don't need to have tests for the logged output. That might make these tests a little too brittle

Reply:
Addressed in https://github.com/HubSpot/hubspot-local-dev-lib/pull/195/commits/e98326c0500fa5716cd9765c2558f402dabeec3d